[LC-163] Sign by the YubiHSM from the block version of 0.3.

hyerim-kim opened the pull request:
- Load key from HSM.
- Sign by HSM.
- Move functions of signing to the `loopchain.crypto.signature` package.

hyerim-kim committed:
[LC-163] Load a key from HSM and sign to a block.

- Edit some classes for signature.
- Add logic about signature for HSM in classes related with Block of version 3.

hyerim-kim committed:
[LC-163] Polish some code and configurations.

yakkle commented on loopchain/configure_default.py:
```diff
@@ -409,6 +409,18 @@ def is_support_node_function(cls, node_function, node_type):
+HSM_ENABLE_USE = False
+HSM_YUBI_CONNECTOR = "http://127.0.0.1:12345"
+HSM_YUBI_SESSION_ID = 1
+HSM_YUBI_SESSION_PASSWORD = "password"
```
default password is `password`?

hyerim-kim replied:
Yes, the default password of the YubiHSM device.

hyerim-kim commented:
retest this please

hyerim-kim committed:
[LC-163] Add the yubihsm package to `requirements.txt`.

hyerim-kim committed:
[LC-163] Move functions of signing to a signer from a block builder.

zsaladin commented:
In my opinion it is good to modify the structure for better abstraction.
The guidance is followed
1. Replace hardcoded signing logic with signer.sign in BlockBuilder
   -  BlockBuilder.peer_private_key -> BlockBuilder.signer.
2. Replace hardcoded signature verification logic with sign_verifier.verify_hash or sign_verifier.verify_data
3. Signer and SignVerifier should be restructured. How about using strategy pattern?

zsaladin commented on loopchain/tools/hsm_helper.py:
```diff
@@ -0,0 +1,83 @@
+        )
+        return key_info['public_key'].native
+
+    def ecdsa_sign(self, message: bytes, is_raw: bool = True, digest=hashlib.sha3_256()):
```
Default parameter will be cached.

hyerim-kim committed:
[LC-163] Move functions of verifying to a signature verifier from a block verifier.

zsaladin commented on loopchain/blockchain/blocks/v0_3/block_builder.py:
```diff
@@ -176,6 +188,28 @@ def _build_hash(self):
         block_prover = BlockProver(leaves, BlockProverType.Block)
         return block_prover.get_proof_root()
 
+    def build_peer_id(self):
```
BlockBuilder will have `signer`. Then `build_peer_id` method does not needed. `signer` has address(id)

zsaladin commented on loopchain/channel/channel_service.py:
```diff
@@ -46,7 +47,7 @@ def __init__(self, channel_name, amqp_target, amqp_key):
         self.__score_container: CommonSubprocess = None
         self.__score_info: dict = None
-        self.__peer_auth: Signer = None
+        self.__peer_auth: SignVerifier = None
```
peer_auth must be `Signer` not `SignVerifier`

hyerim-kim committed:
[LC-163] Divide modules `signer` and `verifier` from `signature.py` and remove the file.

hyerim-kim committed:
[LC-163] Move functions of signing in the Transaction to the `loopchain.crypto.signature` package.

- Move functions of signing in the Transaction to the `loopchain.crypto.signature` package.
- Change the type of address in signers and verifiers from `str` to `loopchain.blockchain.types.Address`.

hyerim-kim committed:
[LC-163] Edit unit test cases.
